numpy/core/tests/test_records.py

```diff
@@ -416,6 +416,21 @@ def test_pickle_3(self):
             assert_(pa.flags.writeable)
             assert_(pa.flags.aligned)
 
+    def test_pickle_void(self):
+        # issue gh-13593
+        dt = np.dtype([('obj', 'O')])
+        a = np.empty(1, dtype=dt)
+        data = (bytearray(b'eman'),)
+        a['obj'] = data
+        state = a[0].__reduce__()
+        # If the data is bytes, it will pickle the address of the content
+        # which cannot be saved/restored
+        assert not isinstance(state[1][1], bytes)
```

Comment:
I don't quite understand the point you're making here.

Reply:
`state` is now `(<built-in function scalar>, (dtype([('obj', 'O')]), (bytearray(b'eman'),)))`. Before it was `(<built-in function scalar>, (dtype([('obj', 'O')]), b'xxxxxxx'))` where `xxxxxxxx` is the address of data. The result of `__reduce__` is passed to `pickle` for storage.

I tried to say that we need to make sure `a[0]` is not stored as a pointer (the address of `data`, rather it should be `a[0]` itself

Reply:
I think then this test would be clearer as
```python
ctor, args = a[0].__reduce__()
# check the contructor is what we expect before interpreting the arguments
assert ctor is np.core.multiarray.scalar
dtype, obj = args
# make sure we did not pickle the address
assert not isinstance(obj, bytes)
```

Reply:
adopted